src/client/Client.h

```diff
@@ -1222,6 +1229,207 @@ class Client : public Dispatcher, public md_config_obs_t {
   struct initialize_state_t initialize_state;
 
 private:
+  class C_Read_Finisher : public Context {
+  public:
+    bool iofinished;
+    void finish_io(int r);
+    void finish_onuninline(int r);
+
+    C_Read_Finisher(Client *clnt, Context *onfinish, Context *onuninline,
+                    Context *iofinish, bool is_read_async, int have,
+                    bool movepos, utime_t start, Fh *f, Inode *in,
+                    uint64_t fpos, int64_t offset, uint64_t size)
+      : clnt(clnt), onfinish(onfinish), onuninline(onuninline),
+        iofinish(iofinish), is_read_async(is_read_async), have(have),
+        f(f), in(in), start(start), fpos(fpos), offset(offset),
+        size(size), movepos(movepos) {
+      iofinished_r = 0;
+      onuninlinefinished_r = 0;
+      iofinished = false;
+      onuninlinefinished = onuninline == nullptr;
+    }
+
+    void finish(int r) override {
+      //
+      delete this;
+    }
+
+  private:
+    Client *clnt;
+    Context *onfinish;
+    Context *onuninline;
+    Context *iofinish;
+    bool is_read_async;
+    int have;
+    Fh *f;
+    Inode *in;
+    utime_t start;
+    uint64_t fpos;
+    int64_t offset;
+    uint64_t size;
+    int64_t iofinished_r;
+    int64_t onuninlinefinished_r;
+    bool onuninlinefinished;
+    bool movepos;
+    bool try_complete();
+  };
+
+  struct CRF_onuninline : public Context {
+    Client::C_Read_Finisher *CRF;
+
+    CRF_onuninline()
+      : CRF(nullptr) {}
+
+    void finish(int r) override {
+      CRF->finish_onuninline(r);
+      complete(r);
```

Comment:
Won't this cause infinitely loop ? The `Context::complete()` will call the `finish()` again.

Reply:
Oops, yes. That may have been a left over from the class deriving from C_SaferCond at one point during my development.